provider/provider_test.go

```diff
@@ -82,6 +89,100 @@ func testProviderUpgrade(t *testing.T, dir string, opts *testProviderUpgradeOpti
+	secondProgram        []byte
+	firstProgramOptions  []opttest.Option
+	secondProgramOptions []opttest.Option
+}
+
+// testProviderCodeChanges tests two different runs of a pulumi program. This allow you to run
+// pulumi up with an initial program, change the code of the program and then run another pulumi command
+func testProviderCodeChanges(t *testing.T, opts *testProviderCodeChangesOptions) *pulumitest.PulumiTest {
+	skipIfShort(t)
+	t.Parallel()
+	t.Helper()
+
+	workdir := t.TempDir()
+	cacheDir := filepath.Join("testdata", "recorded", "TestProviderUpgrade", t.Name())
+	err := os.MkdirAll(cacheDir, 0755)
+	assert.NoError(t, err)
+	stackExportFile := filepath.Join(cacheDir, "stack.json")
+
+	err = os.WriteFile(filepath.Join(workdir, "Pulumi.yaml"), opts.firstProgram, 0o600)
+	require.NoError(t, err)
+
+	options := []opttest.Option{
+		opttest.SkipInstall(),
+		opttest.NewStackOptions(optnewstack.DisableAutoDestroy()),
+	}
+	for _, o := range opts.firstProgramOptions {
+		options = append(options, o)
+	}
+
+	pt := pulumitest.NewPulumiTest(t, workdir, options...)
+
+	var export *apitype.UntypedDeployment
+	export, err = tryReadStackExport(stackExportFile)
+	if err != nil {
+		pt.Up()
+		grptLog := pt.GrpcLog()
+		grpcLogPath := filepath.Join(cacheDir, "grpc.json")
+		pt.T().Logf("writing grpc log to %s", grpcLogPath)
+		grptLog.WriteTo(grpcLogPath)
+
+		e := pt.ExportStack()
+		export = &e
+		err = writeStackExport(stackExportFile, export, true)
+		assert.NoError(t, err)
+	}
+
+	secondOptions := []opttest.Option{
+		opttest.SkipInstall(),
+		opttest.NewStackOptions(optnewstack.EnableAutoDestroy()),
+	}
+	for _, o := range opts.secondProgramOptions {
+		secondOptions = append(secondOptions, o)
+	}
+	err = os.WriteFile(filepath.Join(workdir, "Pulumi.yaml"), opts.secondProgram, 0o600)
+	require.NoError(t, err)
+	secondTest := pulumitest.NewPulumiTest(t, workdir, secondOptions...)
+	secondTest.ImportStack(*export)
+
+	return secondTest
+}
+
+// pulumiUpWithSnapshot will only run the up portion of the test if the plan has changed since the
+// last time the test was run.
+//
+// This should be used when the plan is a good representation of what you are testing. Sometimes
+// there are issues where the plan is consistent, but the apply fails. In those cases a snapshot test is not
+// a good fit.
+func pulumiUpWithSnapshot(t *testing.T, pulumiTest *pulumitest.PulumiTest) {
+	workdir := os.TempDir()
+	cwd, err := os.Getwd()
+	assert.NoError(t, err)
+	cacheDir := filepath.Join(cwd, "testdata", "recorded", "TestProviderUpgrade", t.Name())
+	planFile := filepath.Join(cacheDir, "plan.json")
+	if ok, _ := exists(planFile); ok {
+		err := os.MkdirAll(cacheDir, 0755)
+		assert.NoError(t, err)
+		tmpPlanFile := filepath.Join(workdir, "plan.json")
+
+		pulumiTest.Preview(optpreview.Plan(tmpPlanFile))
+
+		if equal := planEqual(t, planFile, tmpPlanFile); equal {
+			return
+		}
+
+		pulumiTest.T().Log("Plan is not equal, re-running up")
```

Comment:
This is https://www.pulumi.com/docs/concepts/update-plans/ right?

Reply:
Yes!